File: hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncClientScanner.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hadoop.hbase.client;

import static org.apache.hadoop.hbase.HConstants.EMPTY_END_ROW;
import static org.apache.hadoop.hbase.HConstants.EMPTY_START_ROW;
import static org.apache.hadoop.hbase.client.ConnectionUtils.createScanResultCache;
import static org.apache.hadoop.hbase.client.ConnectionUtils.getLocateType;
import static org.apache.hadoop.hbase.client.ConnectionUtils.incRPCCallsMetrics;
import static org.apache.hadoop.hbase.client.ConnectionUtils.incRPCRetriesMetrics;
import static org.apache.hadoop.hbase.client.ConnectionUtils.incRegionCountMetrics;
import static org.apache.hadoop.hbase.client.ConnectionUtils.isRemote;
import static org.apache.hadoop.hbase.client.ConnectionUtils.timelineConsistentRead;
import static org.apache.hadoop.hbase.util.FutureUtils.addListener;

import java.io.IOException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import org.apache.hadoop.hbase.HRegionLocation;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.client.metrics.ScanMetrics;
import org.apache.hadoop.hbase.ipc.HBaseRpcController;
import org.apache.yetus.audience.InterfaceAudience;

import org.apache.hbase.thirdparty.io.netty.util.Timer;

import org.apache.hadoop.hbase.shaded.protobuf.RequestConverter;
import org.apache.hadoop.hbase.shaded.protobuf.generated.ClientProtos.ClientService;
import org.apache.hadoop.hbase.shaded.protobuf.generated.ClientProtos.ClientService.Interface;
import org.apache.hadoop.hbase.shaded.protobuf.generated.ClientProtos.ScanRequest;
import org.apache.hadoop.hbase.shaded.protobuf.generated.ClientProtos.ScanResponse;

/**
 * The asynchronous client scanner implementation.
 * <p>
 * Here we will call OpenScanner first and use the returned scannerId to create a
 * {@link AsyncScanSingleRegionRpcRetryingCaller} to do the real scan operation. The return value of
 * {@link AsyncScanSingleRegionRpcRetryingCaller} will tell us whether open a new scanner or finish
 * scan.
 */
@InterfaceAudience.Private
class AsyncClientScanner {

  // We will use this scan object during the whole scan operation. The
  // AsyncScanSingleRegionRpcRetryingCaller will modify this scan object directly.
  private final Scan scan;

  private final ScanMetrics scanMetrics;

  private final AdvancedScanResultConsumer consumer;

  private final TableName tableName;

  private final AsyncConnectionImpl conn;

  private final Timer retryTimer;

  private final long pauseNs;

  private final long pauseForCQTBENs;

  private final int maxAttempts;

  private final long scanTimeoutNs;

  private final long rpcTimeoutNs;

  private final int startLogErrorsCnt;

  private final ScanResultCache resultCache;

  public AsyncClientScanner(Scan scan, AdvancedScanResultConsumer consumer, TableName tableName,
      AsyncConnectionImpl conn, Timer retryTimer, long pauseNs, long pauseForCQTBENs,
      int maxAttempts, long scanTimeoutNs, long rpcTimeoutNs, int startLogErrorsCnt) {
    if (scan.getStartRow() == null) {
      scan.withStartRow(EMPTY_START_ROW, scan.includeStartRow());
    }
    if (scan.getStopRow() == null) {
      scan.withStopRow(EMPTY_END_ROW, scan.includeStopRow());
    }
    this.scan = scan;
    this.consumer = consumer;
    this.tableName = tableName;
    this.conn = conn;
    this.retryTimer = retryTimer;
    this.pauseNs = pauseNs;
    this.pauseForCQTBENs = pauseForCQTBENs;
    this.maxAttempts = maxAttempts;
    this.scanTimeoutNs = scanTimeoutNs;
    this.rpcTimeoutNs = rpcTimeoutNs;
    this.startLogErrorsCnt = startLogErrorsCnt;
    this.resultCache = createScanResultCache(scan);
    if (scan.isScanMetricsEnabled()) {
      this.scanMetrics = new ScanMetrics();
      consumer.onScanMetricsCreated(scanMetrics);
    } else {
      this.scanMetrics = null;
    }
  }

  private static final class OpenScannerResponse {

    public final HRegionLocation loc;

    public final boolean isRegionServerRemote;

    public final ClientService.Interface stub;

    public final HBaseRpcController controller;

    public final ScanResponse resp;

    public OpenScannerResponse(HRegionLocation loc, boolean isRegionServerRemote, Interface stub,
        HBaseRpcController controller, ScanResponse resp) {
      this.loc = loc;
      this.isRegionServerRemote = isRegionServerRemote;
      this.stub = stub;
      this.controller = controller;
      this.resp = resp;
    }
  }

  private final AtomicInteger openScannerTries = new AtomicInteger();

  private CompletableFuture<OpenScannerResponse> callOpenScanner(HBaseRpcController controller,
      HRegionLocation loc, ClientService.Interface stub) {
    boolean isRegionServerRemote = isRemote(loc.getHostname());
    incRPCCallsMetrics(scanMetrics, isRegionServerRemote);
    if (openScannerTries.getAndIncrement() > 1) {
      incRPCRetriesMetrics(scanMetrics, isRegionServerRemote);
    }
    CompletableFuture<OpenScannerResponse> future = new CompletableFuture<>();
    try {
      ScanRequest request = RequestConverter.buildScanRequest(loc.getRegion().getRegionName(), scan,
        scan.getCaching(), false);
      stub.scan(controller, request, resp -> {
        if (controller.failed()) {
          future.completeExceptionally(controller.getFailed());
          return;
        }
        future.complete(new OpenScannerResponse(loc, isRegionServerRemote, stub, controller, resp));
      });
    } catch (IOException e) {
      future.completeExceptionally(e);
    }
    return future;
  }

  private void startScan(OpenScannerResponse resp) {
    addListener(
      conn.callerFactory.scanSingleRegion().id(resp.resp.getScannerId()).location(resp.loc)
        .remote(resp.isRegionServerRemote)
        .scannerLeaseTimeoutPeriod(resp.resp.getTtl(), TimeUnit.MILLISECONDS).stub(resp.stub)
        .setScan(scan).metrics(scanMetrics).consumer(consumer).resultCache(resultCache)
        .rpcTimeout(rpcTimeoutNs, TimeUnit.NANOSECONDS)
        .scanTimeout(scanTimeoutNs, TimeUnit.NANOSECONDS).pause(pauseNs, TimeUnit.NANOSECONDS)
        .pauseForCQTBE(pauseForCQTBENs, TimeUnit.NANOSECONDS).maxAttempts(maxAttempts)
        .startLogErrorsCnt(startLogErrorsCnt).start(resp.controller, resp.resp),
      (hasMore, error) -> {
        if (error != null) {
          consumer.onError(error);
          return;
        }
        if (hasMore) {
          openScanner();
        } else {
          consumer.onComplete();
        }
      });
  }

  private CompletableFuture<OpenScannerResponse> openScanner(int replicaId) {
    return conn.callerFactory.<OpenScannerResponse> single().table(tableName)
      .row(scan.getStartRow()).replicaId(replicaId).locateType(getLocateType(scan))
      .rpcTimeout(rpcTimeoutNs, TimeUnit.NANOSECONDS)
      .operationTimeout(scanTimeoutNs, TimeUnit.NANOSECONDS).pause(pauseNs, TimeUnit.NANOSECONDS)
      .pauseForCQTBE(pauseForCQTBENs, TimeUnit.NANOSECONDS).maxAttempts(maxAttempts)
      .startLogErrorsCnt(startLogErrorsCnt).action(this::callOpenScanner).call();
  }

  private long getPrimaryTimeoutNs() {
    return TableName.isMetaTableName(tableName) ? conn.connConf.getPrimaryMetaScanTimeoutNs()
      : conn.connConf.getPrimaryScanTimeoutNs();
  }

  private void openScanner() {
    incRegionCountMetrics(scanMetrics);
    openScannerTries.set(1);
    addListener(timelineConsistentRead(conn.getLocator(), tableName, scan, scan.getStartRow(),
      getLocateType(scan), this::openScanner, rpcTimeoutNs, getPrimaryTimeoutNs(), retryTimer,
      conn.getConnectionMetrics()), (resp, error) -> {
        if (error != null) {
          consumer.onError(error);
          return;
        }
        startScan(resp);
      });
  }

  public void start() {
    openScanner();
  }
}

File: hbase-native-client/core/async-client-scanner.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "core/async-client-scanner.h"

#include <algorithm>
#include <iterator>
#include <limits>
#include <stdexcept>

namespace hbase {

AsyncClientScanner::AsyncClientScanner(
    std::shared_ptr<AsyncConnection> conn, std::shared_ptr<Scan> scan,
    std::shared_ptr<pb::TableName> table_name, std::shared_ptr<RawScanResultConsumer> consumer,
    nanoseconds pause, uint32_t max_retries, nanoseconds scan_timeout_nanos,
    nanoseconds rpc_timeout_nanos, uint32_t start_log_errors_count)
    : conn_(conn),
      scan_(scan),
      table_name_(table_name),
      consumer_(consumer),
      pause_(pause),
      max_retries_(max_retries),
      scan_timeout_nanos_(scan_timeout_nanos),
      rpc_timeout_nanos_(rpc_timeout_nanos),
      start_log_errors_count_(start_log_errors_count) {
  results_cache_ = std::make_shared<ScanResultCache>();
  max_attempts_ = ConnectionUtils::Retries2Attempts(max_retries);
}

void AsyncClientScanner::Start() { OpenScanner(); }

folly::Future<std::shared_ptr<OpenScannerResponse>> AsyncClientScanner::CallOpenScanner(
    std::shared_ptr<hbase::RpcClient> rpc_client,
    std::shared_ptr<hbase::HBaseRpcController> controller,
    std::shared_ptr<hbase::RegionLocation> loc) {
  open_scanner_tries_++;

  auto preq = RequestConverter::ToScanRequest(*scan_, loc->region_name(), scan_->Caching(), false);

  auto self(shared_from_this());
  VLOG(5) << "Calling RPC Client to open the scanner for region:" << loc->DebugString();
  return rpc_client
      ->AsyncCall(loc->server_name().host_name(), loc->server_name().port(), std::move(preq),
                  security::User::defaultUser(), "ClientService")
      .then([self, loc, controller, rpc_client](const std::unique_ptr<Response>& presp) {
        VLOG(5) << "Scan Response:" << presp->DebugString();
        return std::make_shared<OpenScannerResponse>(rpc_client, presp, loc, controller);
      });
}

void AsyncClientScanner::OpenScanner() {
  auto self(shared_from_this());
  open_scanner_tries_ = 1;

  auto caller = conn_->caller_factory()
                    ->Single<std::shared_ptr<OpenScannerResponse>>()
                    ->table(table_name_)
                    ->row(scan_->StartRow())
                    ->locate_type(GetLocateType(*scan_))
                    ->rpc_timeout(rpc_timeout_nanos_)
                    ->operation_timeout(scan_timeout_nanos_)
                    ->pause(pause_)
                    ->max_retries(max_retries_)
                    ->start_log_errors_count(start_log_errors_count_)
                    ->action([&](std::shared_ptr<hbase::HBaseRpcController> controller,
                                 std::shared_ptr<hbase::RegionLocation> loc,
                                 std::shared_ptr<hbase::RpcClient> rpc_client)
                                 -> folly::Future<std::shared_ptr<OpenScannerResponse>> {
                                   return CallOpenScanner(rpc_client, controller, loc);
                                 })
                    ->Build();

  caller->Call()
      .then([this, self](std::shared_ptr<OpenScannerResponse> resp) {
        VLOG(3) << "Opened scanner with id:" << resp->scan_resp_->scanner_id()
                << ", region:" << resp->region_location_->DebugString() << ", starting scan";
        StartScan(resp);
      })
      .onError([this, self](const folly::exception_wrapper& e) {
        VLOG(3) << "Open scan request received error:" << e.what();
        consumer_->OnError(e);
      })
      .then([caller, self](const auto r) { return r; });
}

void AsyncClientScanner::StartScan(std::shared_ptr<OpenScannerResponse> resp) {
  auto self(shared_from_this());
  auto caller = conn_->caller_factory()
                    ->Scan()
                    ->scanner_id(resp->scan_resp_->scanner_id())
                    ->region_location(resp->region_location_)
                    ->scanner_lease_timeout(TimeUtil::MillisToNanos(resp->scan_resp_->ttl()))
                    ->scan(scan_)
                    ->rpc_client(resp->rpc_client_)
                    ->consumer(consumer_)
                    ->results_cache(results_cache_)
                    ->rpc_timeout(rpc_timeout_nanos_)
                    ->scan_timeout(scan_timeout_nanos_)
                    ->pause(pause_)
                    ->max_retries(max_retries_)
                    ->start_log_errors_count(start_log_errors_count_)
                    ->Build();

  caller->Start(resp->controller_, resp->scan_resp_, resp->cell_scanner_)
      .then([caller, self](const bool has_more) {
        if (has_more) {
          // open the next scanner on the next region.
          self->OpenScanner();
        } else {
          self->consumer_->OnComplete();
        }
      })
      .onError([caller, self](const folly::exception_wrapper& e) { self->consumer_->OnError(e); })
      .then([caller, self](const auto r) { return r; });
}

RegionLocateType AsyncClientScanner::GetLocateType(const Scan& scan) {
  // TODO: In C++, there is no Scan::IncludeStartRow() and Scan::IncludeStopRow().
  // When added, this method should be modified to return other RegionLocateTypes
  // (see ConnectionUtils.java #getLocateType())
  // TODO: When reversed scans are implemented, return other RegionLocateTypes
  return RegionLocateType::kCurrent;
}

}  // namespace hbase
